packages/relay-runtime/network/RelayNetwork.js

```diff
@@ -43,32 +43,45 @@ function create(fetch: FetchFunction, subscribe?: SubscribeFunction): Network {
     operation: ConcreteBatch,
     variables: Variables,
     cacheConfig?: ?CacheConfig,
-    uploadables?: UploadableMap,
-  ): SyncOrPromise<RelayResponsePayload> {
-    const onSuccess = payload =>
-      normalizePayload(operation, variables, payload);
+    uploadables?: ?UploadableMap,
+    observer: SingleObserver<RelayResponsePayload>,
+  ): Disposable {
+    let isDisposed = false;
+    const onCompleted = payload => {
+      if (isDisposed) {
+        return;
+      }
+      let relayPayload;
+      try {
+        relayPayload = normalizePayload(operation, variables, payload);
+      } catch (err) {
+        observer.onError(err);
+        return;
+      }
+      observer.onCompleted(relayPayload);
+    };
+    const onError = error => {
+      if (isDisposed) {
+        return;
+      }
+      observer.onError(error);
+    };
     const response = fetch(operation, variables, cacheConfig, uploadables);
-    switch (response.kind) {
-      case 'promise':
-        return {
-          kind: 'promise',
-          promise: response.promise.then(onSuccess),
-        };
-      case 'data':
-        return {
-          kind: 'data',
-          data: (response.data && onSuccess(response.data)) || null,
-        };
-      case 'error':
-        return response;
-      default:
-        (response.kind: empty);
-        invariant(
-          false,
-          'RelayNetwork: unsupported response kind `%s`',
-          response.kind,
-        );
+    invariant(
+      typeof response === 'object' && response !== null,
+      'RelayNetwork: Expected fetch function to return an object, got `%s`.',
+      response,
+    );
+    if (typeof response.then === 'function') {
+      response.then(onCompleted).catch(onError);
```

Comment:
should this be `.then(onCompleted, onError)`? i think that matches what was there before. otherwise if `observer.onCompleted` somehow throws, you'd end up calling `observer.onError`, which i don't think is right